harvard-lts/drs-import-management-service
# Interaction with the actual MQs

**Interaction with the actual MQs**
* * *

**GitHub Issue**: ([link](https://github.com/harvard-lts/HDC/issues/106))

* Other Relevant Links: [API & Queue Messaging Definitions](https://docs.google.com/document/d/1Q3KqCRlmcyrQfQxfCYWLUAKSs9DTOj9yp3m4ovrRA7w)

# What does this Pull Request do?
The message publishing and listening logics are evolved to interact with the real MQs. Each MQ (Transfer and Process) has its own publisher and its own listener.

Publishing and listening operations have been added for the entire ingest process. New methods have been added to the IngestService to orchestrate and manage each step of the ingest process.

Persistence logics have been added, but only at the domain level. The MongoDB implementation is currently dummy and will be implemented at [HDC-104](https://github.com/harvard-lts/HDC/issues/104)

There are many TODOs related to exception and error handling for ingest operations. They will be covered in a new specific issue related to error handling ([HDC-124](https://github.com/harvard-lts/HDC/issues/124)).

As there is a lot of common logic for the publishers and listeners of the different MQs, it will be perceived that a major refactor has been added to eliminate duplication and reuse code.

# How should this be tested?

A description of what steps someone could take to:
* Create/Update .env and .test.env files as described in README.md
* Run the local development environment: `docker-compose -f docker-compose-dev.yml up`
* Access local development ActiveMQ frontend http://localhost:8161/admin/ (User: admin, Password: admin)
* Go to queues section: http://localhost:8161/admin/queues.jsp
* POST to http://localhost:10580/ingest
* In the queues section, the number of queued messages for transfer-ready queue should increase by 1. To check enqueued message body click the queue name and then click the message id.
* Now, within queues section, click "Send To" option for queue dropbox-transfer-status, to simulate that a dropbox transfer has been done and a message has been enqueued from Starfish. For message body, type: {"package_id": "dummy_package_id", "transfer_status": "successful", "destination_path": "dummy_destination_path"} and click send.
* In the queues section, the number of messages queued for dims-data-ready queue should be increased by 1. You can verify that the message body is correct in the same way as for transfer-ready queue.
* Now, within the queues section, select the "Send To" option for drs-ingest-status queue, this time to simulate that the translation service has notified the ingest process completion. For the body of the message, type: {"package_id": "dummy_package_id", "application_name": "Dataverse", "batch_ingest_status": "successful", "message":"dummy"}
* In te queues section, check that the message has been dequeued.
* Run the local test environment: `docker-compose -f docker-compose-test.yml up`
* Execute tests: `docker exec -it test_runner pytest test`

# Test coverage
Yes/No: Are changes in this pull-request covered by:
- unit tests: Yes
- integration tests: Yes

# Interested parties
@ives1227 @jessjass 

Keyword: resolved
